c2rust-transpile/src/c_ast/mod.rs

```diff
@@ -486,6 +492,71 @@ impl TypedAstContext {
         self.index(resolved_typ_id)
     }
 
+    /// Extract decl of referenced function.
+    /// Looks for ImplicitCast(FunctionToPointerDecay, DeclRef(function_decl))
+    pub fn function_declref_decl(&self, func_expr: CExprId) -> Option<&CDeclKind> {
+        use CastKind::FunctionToPointerDecay;
+        if let CExprKind::ImplicitCast(_, fexp, FunctionToPointerDecay, _, _) = self[func_expr].kind
+        {
+            if let CExprKind::DeclRef(_ty, decl_id, _rv) = &self[fexp].kind {
+                let decl = &self.index(*decl_id).kind;
+                assert!(matches!(decl, CDeclKind::Function { .. }));
+                return Some(decl);
+            }
+        }
+        None
+    }
+
+    /// Return the list of types for a list of declared function parameters.
+    pub fn tys_of_params(&self, parameters: &[CDeclId]) -> Option<Vec<CQualTypeId>> {
+        let mut param_tys = vec![];
+        for p in parameters {
+            match self.index(*p).kind {
+                CDeclKind::Variable { typ, .. } => param_tys.push(CQualTypeId::new(typ.ctype)),
+                _ => return None,
```

Comment:
Does this ever fail?  What does it mean if it does?  One call is just `.unwrap()`ed and the other doesn't seem to ever be `None` in our tests.

If we're not sure yet how to handle this case, it'd probably be better to panic (like the `assert!` in `fn function_declref_decl`.